clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java

```diff
@@ -118,11 +193,41 @@ public AddPartitionsToTxnRequestData data() {
 
     @Override
     public AddPartitionsToTxnResponse getErrorResponse(int throttleTimeMs, Throwable e) {
-        final HashMap<TopicPartition, Errors> errors = new HashMap<>();
-        for (TopicPartition partition : partitions()) {
-            errors.put(partition, Errors.forException(e));
+        Errors error = Errors.forException(e);
+        if (version < 4) {
+            final HashMap<TopicPartition, Errors> errors = new HashMap<>();
+            for (TopicPartition partition : partitions()) {
+                errors.put(partition, error);
+            }
+            return new AddPartitionsToTxnResponse(throttleTimeMs, errors);
+        } else {
+            AddPartitionsToTxnResponseData response = new AddPartitionsToTxnResponseData();
```

Comment:
The logic in here makes me wonder if we should add a top-level error code in the response.

Reply:
There is 1 error we can return as top level errors if that helps. The rest are topic or partition level.

Reply:
Many new APIs do include a top-level error code, so there is definitely precedent. I kind of like giving the broker an easy way to indicate a failure without requiring it to do a bunch of wasteful work. It's not a common case though, so I'm not sure how much it matters.